### PR TITLE
sourceMap.content "auto" option

### PR DESCRIPTION
**Pull request by Hoten**
I originally made these changes on UglifyJS2, but I merged them to Terser in the hopes they'd get accepted here :)

My original PR on UglifyJS: https://github.com/mishoo/UglifyJS2/pull/3220

Also involved adding these changes: https://github.com/mishoo/UglifyJS2/pull/3058

# overview
Implements https://github.com/mishoo/UglifyJS2/issues/3219

```
--source-map "content=auto"
```

Allows for easier composition of source maps.

If an input file specifies a sourceMappingURL comment, the source map will be fetched.
If there is no comment, given an input file of `src/file.js`, the map will be looked for
at `src/file.js.map` and `src/file.map`. Maps can fetched via HTTP too.

```
--source-map "content=auto,contents=file1.js*file1.js.map|file2.js*path/to/map.js.map|..."
```

Allows for overriding the auto resolution strategy.

# other stuff

Open to a different format for `sourceMap.contents`.

This is a super set of the other source map content options. I suggest removing those in a new major version, and making this "auto" behavior the default.

### CODE REVIEW

File: bin/uglifyjs [hunk 1]
```diff
@@ -221,6 +223,59 @@ function run() {
     } catch (ex) {
         fatal(ex);
     }
+    if (content == "auto") {
+        var to_ascii = function(b64) {
```
Comment (L2jLiga) [inline bin/uglifyjs]: I think we can use `UglifyJS.to_ascii`

File: bin/uglifyjs [hunk 2]
```diff
@@ -221,6 +223,59 @@ function run() {
     } catch (ex) {
         fatal(ex);
     }
+    if (content == "auto") {
```
Comment (L2jLiga) [inline bin/uglifyjs]: I would like to use strict equal

File: bin/uglifyjs [hunk 3]
```diff
@@ -221,6 +223,59 @@ function run() {
...
+                if (fs.existsSync(name + '.map')) {
+                    return read_file(name + '.map');
+                }
+                if (/\.js$/.test(name) && fs.existsSync(name.slice(0, -2) + 'map')) {
+                    return read_file(name.slice(0, -2) + 'map');
+                }
+                return;
+            }
+
+            if (sourceMappingURL.indexOf('data:application') == 0) {
```
Comment (L2jLiga) [inline bin/uglifyjs]: `sourceMappingURL.includes('data:application')` ? 🙃

File: bin/uglifyjs [hunk 4]
```diff
@@ -221,6 +223,59 @@ function run() {
...
+                var match = /data:application\/json(;.*?)?;base64,(.*)/.exec(sourceMappingURL);
+                return to_ascii(match[2]);
+            }
+
+            if (sourceMappingURL.indexOf('http') == 0) {
+                try {
+                    return execSync('curl -s ' + sourceMappingURL).toString();
```
Comment (L2jLiga) [inline bin/uglifyjs]: On my windows machine curl is not installed.... 😖

File: lib/minify.js [hunk 5]
```diff
@@ -157,13 +166,28 @@ function minify(files, options) {
             }
             options.parse = options.parse || {};
             options.parse.toplevel = null;
+            var source_map_content = options.sourceMap && options.sourceMap.content;
+            if (typeof source_map_content == "string" && source_map_content != "inline" && source_map_content != "auto") {
+                source_map_content = parse_source_map(source_map_content);
+            }
+            source_maps = source_map_content && Object.create(null);
```
Comment (L2jLiga) [inline lib/minify.js]: According to https://github.com/terser-js/terser/pull/300#issuecomment-469037773 I guess it would be much better to use smth like:
```javascript
if (source_map_content) source_maps = new Set();
```

Comment (L2jLiga) [reply lib/minify.js]: At least it will reduce variety in code which doing same things

File: bin/uglifyjs [hunk 6]
```diff
@@ -221,6 +223,59 @@ function run() {
...
+            return new Buffer(b64, "base64").toString();
+        };
+
+        var resolve_source_map_content = function(name, code) {
+            var sourceMappingURLMatch = /\n\/\/# sourceMappingURL=(.*)/.exec(code);
+            var sourceMappingURL = sourceMappingURLMatch ? sourceMappingURLMatch[1] : null;
+            
+            if (!sourceMappingURL) {
+                // infer
+                if (fs.existsSync(name + '.map')) {
```
Comment (L2jLiga) [inline bin/uglifyjs]: ESLint: Strings must use doublequote. (quotes)

File: lib/sourcemap.js [hunk 7]
```diff
@@ -62,20 +62,24 @@ function SourceMap(options) {
...
     });
-    var orig_map = options.orig && new MOZ_SourceMap.SourceMapConsumer(options.orig);
-
-    if (orig_map) {
-        orig_map.sources.forEach(function(source) {
-            var sourceContent = orig_map.sourceContentFor(source, true);
-            if (sourceContent) {
-                generator.setSourceContent(source, sourceContent);
-            }
+    var maps = options.orig && Object.create(null);
```
Comment (L2jLiga) [inline lib/sourcemap.js]: It looks like `Set`

File: bin/uglifyjs [hunk 8]
```diff
@@ -135,7 +136,7 @@ if (program.output == "ast") {
 if (program.parse) {
     if (!program.parse.acorn && !program.parse.spidermonkey) {
         options.parse = program.parse;
-    } else if (program.sourceMap && program.sourceMap.content == "inline") {
+    } else if (content == "inline") {
```
Comment (L2jLiga) [inline bin/uglifyjs]: I'm little bit worried about this changes..
`ESLint: 'content' is not defined. (no-undef)`

File: bin/uglifyjs (hunk 1 above)
Comment (Hoten) [reply bin/uglifyjs]: Done!

File: bin/uglifyjs (hunk 2 above)
Comment (Hoten) [reply bin/uglifyjs]: 👍

File: lib/minify.js (hunk 5 above)
Comment (Hoten) [reply lib/minify.js]: 👍 , but it's actually a Map.

File: bin/uglifyjs (hunk 6 above)
Comment (Hoten) [reply bin/uglifyjs]: 👍

File: lib/sourcemap.js (hunk 7 above)
Comment (Hoten) [reply lib/sourcemap.js]: 👍 , but it's actually a Map.

File: bin/uglifyjs (hunk 8 above)
Comment (Hoten) [reply bin/uglifyjs]: Oops!

File: bin/uglifyjs (hunk 4 above)
Comment (Hoten) [reply bin/uglifyjs]: Replace with the http libs. Required some promises.

File: bin/uglifyjs [hunk 9]
```diff
@@ -4,7 +4,7 @@
 
 "use strict";
 
-require("../tools/exit.js");
+// require("../tools/exit.js");
```
Comment (Hoten) [inline bin/uglifyjs]: I had to remove this to get the tests that expect errors to work. I can't reason about this code or how to augment it - ideas?

**Comment by Hoten** [comment]
> This is a super set of the other source map content options. I suggest removing those in a new major version, and making this "auto" behavior the default.

Should this go for 4.0 or 5.0?

**Comment by Hoten** [comment]
Forgot node 6 doesn't do async/await :( I'll update tomorrow.

File: lib/sourcemap.js (hunk 7 above)
Comment (L2jLiga) [reply lib/sourcemap.js]: My bad)

**Comment by L2jLiga** [comment]
I think it can be part of Terser 4.0

**Comment by Hoten** [comment]
> I think it can be part of Terser 4.0

OK. I'll see about removing the other "modes" in a follow up PR.

**Comment by Hoten** [comment]
OK- removed async/await. Obviously made the code less readable. More proof that async/await is a game changer :)

File: bin/uglifyjs [hunk 10]
```diff
@@ -221,90 +292,99 @@ function run() {
...
-                    col = limit;
+    // TODO: replace with async/await when node 6 support is dropped.
+    var resolveSourcemapsPromise;
+    if (content === "auto") {
+        resolveSourcemapsPromise = resolve_sourcemaps();
+    } else {
+        resolveSourcemapsPromise = Promise.resolve();
+    }
+    return resolveSourcemapsPromise.then(() => {
+        var result = UglifyJS.minify(files, options);
```
Comment (Hoten) [inline bin/uglifyjs]: all of this below is just indentation

**Comment by Hoten** [comment]
What to do about https://github.com/terser-js/terser/pull/298#discussion_r261928693 ?

**Comment by L2jLiga** [comment]
I don't know, I guess it's a hack for old libUV and it can be deleted, not sure..

@fabiosantoscode, @kzc could you clarify?

**Comment by fabiosantoscode** [comment]
@Hoten @L2jLiga node isn't guaranteed to output to stdout/stderr properly if you don't do that workaround. The workaround for *that* is to change instances of process.exit(errCod) with `process.exitCode = errCod` and change the code flow as necessary.

**Comment by fabiosantoscode** [comment]
See the code example reading "this is an example of what not to do" https://nodejs.org/api/process.html#process_process_exit_code

**Comment by Hoten** [comment]
Interesting, never ran into that before. I tweaked the exit workaround code so it'll also work within a promise context, without causing an unresolved rejection (the `throw` was problematic).

**Comment by Hoten** [comment]
Closing / Opening to force travis rerun. They had issues when this last ran.

**Comment by Hoten** [comment]
```
'istanbul' is not recognized as an internal or external command
```

OK, travis env is borked.

**Comment by L2jLiga** [comment]
@Hoten, you can tweak "test" script to exclude Istanbul call and make build green

**Comment by L2jLiga** [comment]
@Hoten there're conflicts after #300 and #288

**Comment by Hoten** [comment]
Resolved conflicts.

**Comment by Hoten** [comment]
Good to merge?

**Comment by fabiosantoscode** [comment]
I've just looked through the PR. It has some issues like generated code in test files, the fact that the `exit()` function is synchronous or asynchronous depending on the situation, regex-based reading of things you wouldn't normally use a regex for...

And I was adding comments one by one to outline them, but I realized the PR can't be accepted because of this:

https://github.com/terser-js/terser/pull/298#discussion_r261928587

If we're going to add this feature, it makes no sense to add it only in the CLI. The maintenance cost is too high and most people are using Terser behind another tool anyway. We can't add it to the CLI and the API at the same time, given the fact that the API is synchronous (IE Terser.minify returns an object, not a Promise). Making the Terser API asynchronous would have a huge impact on Terser users, making them use promises in scripts that might have always been synchronous. I can't push this upon them.

In conclusion, this is an amazing idea in theory, but without a real option to do a synchronous HTTP request present in node core, in practice it doesn't work.

**Comment by Hoten** [comment]
Let's not close this just yet :) Your issues aren't unresolvable.

> If we're going to add this feature, it makes no sense to add it only in the CLI.

Resolving source maps when not explicitly provided by poking around the environment (either the local filesystem or by reaching out to the web) only makes sense in the CLI. AFAICT, `bin/uglifyjs` is free to be as node-y as it wishes, while everything in `lib` (the API) must be shippable to a web context.

Granted, the name `auto` makes less sense in isolation for the API (it's really looking up source maps in a provided map of them: `options.sourceMap.contents`) so perhaps `auto` would be better named `provided` (while still being automatically resolved from the CLI entry).

> The maintenance cost is too high and most people are using Terser behind another tool anyway.

I disagree about the maintenance cost being high, but could you explain further?

The fact that Terser does not handle multiple source maps from multiple inputs is the exact reason this PR exists. Having this tool in my asset pipeline (well, it was uglifyjs, but still) made proper usage of source maps impossible. Good source map support is becoming very important, and Terser shouldn't ignore that.

> We can't add it to the CLI and the API at the same time, given the fact that the API is synchronous (IE Terser.minify returns an object, not a Promise).

I'm not seeing the issue here. Terser.minify still returns an object in this PR. The API is still synchronous - the CLI entry point is not, but that is not the API, `lib` is. 

> Making the Terser API asynchronous would have a huge impact on Terser users, making them use promises in scripts that might have always been synchronous. I can't push this upon them.

I agree, `lib` should remain synchronous. But let's not stick to such standard for `bin`, an entry point no one should be relying on as a library, except via `node bin/uglifyjs ...`. 

> In conclusion, this is an amazing idea in theory, but without a real option to do a synchronous HTTP request present in node core, in practice it doesn't work.

Thanks. But in practice, I think it will. Let's discuss further.

**Comment by connorjclark** [comment]
Bumping b/c I believe I addressed your concerns in my previous comment, and believe this to be a very useful feature.

**Comment by connorjclark** [comment]
gentle ping

**Comment by connorjclark** [comment]
Please consider this patch again and my comments addressing your concerns.
